Off-chain workers docs (#379)
substrate-developer-hub/substrate-developer-hub.github.io: docs/development/module/off-chain-workers.md

---
title: Off-Chain Workers
---

> You can learn about the high level concept of off-chain workers in our [Conceptual Guide](conceptual/core/off-chain-workers.md).

Reviewer: ```suggestion
This article covers the technical aspects of using off-chain workers in a Substrate runtime. For a conceptual overview of off-chain workers see [Conceptual Guide](conceptual/core/off-chain-workers.md).
```

## Using Off-Chain Workers in Runtime

To use off-chain workers in your runtime, include these following modules

> **runtime/src/my_runtime.rs**

Reviewer: For tutorials I like the idea of providing a specific path to the file being modified. I'm not sure it makes sense here since we don't know the name of the reader's pallet or the filesystem location where it is being developed. It may be in its own rust module rather than right in the runtime.

Author: True. But the filename provide the context where the code to be added. I was thinking,

1. skipping the filename altogether - then readers may not know where to add the code. Since some code are added in `lib.rs`, and some in the `runtime.rs`

2. `runtime.rs` (or `pallet.rs`) - this is the best name I can thinking of without the file path

3. `runtime/src/my_runtime.rs` - giving the path, hopefully it gives more clue on where the code to be added

What do you think?

Reviewer: Yeah, I agree. Leaving some kind of path is better than no path.

I don't like the name `runtime.rs` for the pallet though because the complete runtime actually lives in `lib.rs`

Author: True. Actually I have changed all file paths to `runtime/src/offchain_pallet.rs`.

Refer to this commit: 
https://github.com/substrate-developer-hub/substrate-developer-hub.github.io/pull/379/commits/90316e7925704c376a1012fdcb874a87280bfab2

See if the PR is good for approval. Thanks.

```rust
//For better debugging (printout) support
use support::{ debug, dispatch };
use system::offchain;
use sp_runtime::transaction_validity::{
  TransactionValidity, TransactionLongevity, ValidTransaction, InvalidTransaction
};
```

> Include the following associated type in your runtime trait for sending signed and unsigned transactions from off-chain worker.

Reviewer: ```suggestion
Include the following associated types in your pallet's configuration trait for sending signed and unsigned transactions from an off-chain worker.
```

**runtime/src/my_runtime.rs**

```rust
pub trait Trait: timestamp::Trait + system::Trait {
  /// The overarching event type.
  type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
  type Call: From<Call<Self>>;

  type SubmitSignedTransaction: offchain::SubmitSignedTransaction<Self, <Self as Trait>::Call>;
  type SubmitUnsignedTransaction: offchain::SubmitUnsignedTransaction<Self, <Self as Trait>::Call>;
}
```

> Now inside the `decl_module!` dispatch functions section, define the following function and this serves as the entry point of the off-chain worker after every block import.

Reviewer: ```suggestion
Now inside the `decl_module!` section, define the `offchain_worker` function. This function serves as the entry point of the off-chain worker and runs after every block import.
```

**runtime/src/my_runtime.rs**

```rust
decl_module! {
  pub struct Module<T: Trait> for enum Call where origin: T::Origin {
    // ...
    // ...
    fn offchain_worker(block: T::BlockNumber) {
      debug::info!("Hello World.");
    }
  }
}
```

> By default, the offchain worker doesn't have direct access to user keys (even in development environment), but can only access  app-specific subkeys for security reason. You need to define the `KeyTypeId` at the top of your runtime that is used to group your app-specific subkeys as followed.

Reviewer: ```suggestion
By default, the offchain worker doesn't have direct access to user keys (even in development environment), but can only access app-specific subkeys for security reason. You need to define the `KeyTypeId` at the top of your runtime that is used to group your app-specific subkeys as followed.
```

**runtime/src/my_runtime.rs**

```rust
// The key type ID can be any 4-character long string
pub const KEY_TYPE: KeyTypeId = KeyTypeId(*b"abcd");
// ...
// ...
pub mod crypto {
  pub use super::KEY_TYPE;
  use sp_runtime::app_crypto::{app_crypto, sr25519};
  app_crypto!(sr25519, KEY_TYPE);
}
```

> To support this runtime, implement in your library module

Reviewer: ```suggestion
As with any other pallet, your runtime must implement the pallet's configuration trait.
```

This is starting to feel more like a tutorial or recipe.

**runtime/src/lib.rs**

```rust
/// Define the Transaction signer using the Key definition
type SubmitTransaction = system::offchain::TransactionSubmitter<
  my_runtime::crypto::Public, Runtime, UncheckedExtrinsic>;

> impl runtime::Trait for Runtime {

Reviewer: The pallet should not be called `runtime`. How about something like `my_pallet` or `offchain_pallet`?

Author: Using `offchain_pallet` now.

      system::CheckNonce::<Runtime>::from(index),
      system::CheckWeight::<Runtime>::new(),
      transaction_payment::ChargeTransactionPayment::<Runtime>::from(tip),
    );
    let raw_payload = SignedPayload::new(call, extra).ok()?;
    let signature = TSigner::sign(public, &raw_payload)?;
    let address = Indices::unlookup(account);
    let (call, extra, _) = raw_payload.deconstruct();
    Some((call, (address, signature, extra)))
  }
}
```

Inside the `contruct_runtime!` macro where you put all various packages as your runtime enum, add an additional parameter `ValidateUnsigned` if you are using unsigned transactions in off-chain workers.

**runtime/src/lib.rs**

```rust
construct_runtime!(
  pub enum Runtime where
    Block = Block,
    NodeBlock = opaque::Block,
    UncheckedExtrinsic = UncheckedExtrinsic
  {
    // ...
    // ...
    // For using unsigned transactions
    MyRuntime: my_runtime::{ Module, Call, Storage, Event<T>, transaction_validity::ValidateUnsigned }

    // For just using signed transactions, it can just well be:
    // MyRuntime: my_runtime::{ Module, Call, Storage, Event<T> }
  }
);
```

## Insert Keys in `service.rs`

> We have specify a local keystore with `KeyTypeId` to store app-specific sub-keys that are accessible by off-chain workers for signing transactions. But currently there is no sub-keys inside. We can add it in one of the following two ways.

Reviewer: ```suggestion
We have specified a local keystore with `KeyTypeId` to store app-specific keys that are accessible by the off-chain worker for signing transactions. But currently there are no keys inside. We can add keys in one of the following two ways.
```

### Development: Add the First User Key as App Subkey

In development environment, you can add the first user key as the app sub-key. Update the `service.rs` as followed.

**node/src/service.rs**

```rust
pub fn new_full<C: Send + Default + 'static>(config: Configuration<C, GenesisConfig>)
  -> Result<impl AbstractService, ServiceError>
{
  // ...

>   // ...

Reviewer: // --snip--

I believe we've been using this `// --snip--` syntax to follow the rust book.

  // Add this line
  let dev_seed = config.dev_key_seed.clone();

  // ...
  // ...
  let service = builder.with_network_protocol(|_| Ok(NodeProtocol::new()))?
    .with_finality_proof_provider(|client, backend|
      Ok(Arc::new(GrandpaFinalityProofProvider::new(backend, client)) as _)
    )?
    .build()?;

  // Add the following section
  if let Some(seed) = dev_seed {
    service
      .keystore()
      .write()
      .insert_ephemeral_from_seed_by_type::<runtime::my_runtime::crypto::Pair>(
        &seed,
        runtime::my_runtime::KEY_TYPE,
      )
      .expect("Dev Seed should always succeed.");
  }

> }

Reviewer: What key or keys actually get added here? Just Alice? Alice, Bob, etc? Something else?

Author: Referring to: https://github.com/paritytech/substrate/blob/master/client/cli/src/lib.rs#L901-L908

In development, it is only Alice.

```

Then you would be able to sign transactions, and this is good for **DEVELOPMENT only**.

### Add an App Subkey via CLI

> In a more realistic setting, after setting up your substrate node, you can add a new app subkey via

Reviewer: ```suggestion
In a more realistic setting, after setting up your Substrate node, you can add a new app subkey via
```

> the command line interface as followed.

Reviewer: ```suggestion
the command line interface as follows.
```

```bash
# Generate a new account
$ subkey -s generate

# Submit a new key via RPC

> $ http localhost:9933 jsonrpc=2.0 id=1 method=author_insertKey params:='["<YourKeyTypeId>", "<YourSeedPhrase>", "YourPublicKey"]'

Reviewer: My debian system does not come with `http` installed. Maybe `curl` is more popular? Or this can be done through Apps UI. I wonder if we should have a separate doc on making RPC calls?

This comment is not a blocker. I'm just thinking out loud and I think competent developers will be able to take what you've written and figure out how to make the RPC call.

Author: Doing a simple doc/guide on RPC call will be great.

Just realized using `curl` is quite cumbersome, see the update. But this should be more widely-runnable to readers.

```

A new key is now added in the local keystore.

## Signed Transactions

Now you are ready to to make a signed transaction from the off-chain worker as shown in the following.

**runtime/src/my_runtime.rs**

```rust
decl_module! {
  pub struct Module<T: Trait> for enum Call where origin: T::Origin {
    // ...
    // ...
    // This is your regular on-chain extrinsics
    pub fn onchain_callback(origin, _block: T::BlockNumber, input: Vec<u8>) -> dispatch::Result {
      let who = ensure_signed(origin)?;
      debug::info!("{:?}", core::str::from_utf8(&input).unwrap());
      Ok(())
    }

    fn offchain_worker(block: T::BlockNumber) {
      // Here we specify the function to be callback on-chain in next block import
      let call = Call::onchain_callback(block, b"hello world!".to_vec());
      T::SubmitSignedTransaction::submit_signed(call);
    }
  }
}
```

Here, after having defined the on-chain callback function, in the off-chain worker you can specify
that function to be called back in the next block import phase. You then submit a signed transaction
to the node.

> If you look at the implementation of `fn system::offchain::submit_signed ()` in [Substrate codebase](https://github.com/paritytech/substrate), you would realized it is calling the on-chain callback for each of the key in local keystore.

Reviewer: This link should go to a specific place, not just the main Substrate repo.
```suggestion
If you look at the implementation of `fn system::offchain::submit_signed ()` in the [Substrate codebase](https://github.com/paritytech/substrate), you will see it is calling the on-chain callback for each of the key in local keystore.
```

But since you only have one key in the local keystore now, you are calling the function only once back on-chain.

## Unsigned Transactions

With the following code, you are able to send an unsigned transaction back on-chain.

**runtime/src/my_runtime.rs**

```rust
decl_module! {
  pub struct Module<T: Trait> for enum Call where origin: T::Origin {
    // ...
    // ...
    // This is your regular on-chain extrinsics
    pub fn onchain_callback(_origin, _block: T::BlockNumber, input: Vec<u8>) -> dispatch::Result {
      debug::info!("{:?}", core::str::from_utf8(&input).unwrap());
      Ok(())
    }

    fn offchain_worker(block: T::BlockNumber) {
      // Here we specify the function to be callback on-chain in next block import
      let call = Call::onchain_callback(block, b"hello world!".to_vec());
      T::SubmitUnsignedTransaction::submit_unsigned(call);
    }
  }
}
```

By default, all unsigned transactions are treated as invalid transactions. You need to add the following

> code piece in `my_runtime.rs` to whitelist some on-chain functions that are allowed to be called with

Reviewer: ```suggestion
code piece in `my_runtime.rs` to explicitly allow some on-chain functions to be called with
```

Stealing this suggestion from Fuchsia docs https://fuchsia.dev/fuchsia-src/contribute/best-practices/respectful_code#what_are_examples_of_terminology_to_be_avoided

unsigned transaction.

**runtime/src/my_runtime.rs**

```rust
decl_module! {
  // ...
  // ...
}

impl<T: Trait> Module<T> {
  // ...
  // ...
}

#[allow(deprecated)]
impl<T: Trait> support::unsigned::ValidateUnsigned for Module<T> {
  type Call = Call<T>;

  fn validate_unsigned(call: &Self::Call) -> TransactionValidity {

    match call {
      Call::onchain_callback(block, input) => Ok(ValidTransaction {
        priority: 0,
        requires: vec![],
        provides: vec![(block, input).encode()],
        longevity: TransactionLongevity::max_value(),
        propagate: true,
      }),
      _ => InvalidTransaction::Call.into()
    }
  }
}
```

You see that we add a `deprecated` attribute to prevent warning message from display. This is because
this part of the API is still in transition and will be updated in coming Substrate release.

## Parameters in On-Chain Callbacks

When making a on-chain callback, our implementation hashes the function name together with all parameter

> values and store it to be called in next import. If we find that the hash value existed, meaning a function with the same parameter value set has been called before, for signed transaction this causes a function replacement if this call is set with a higher priority; and for unsigned transaction, this callback is simply ignored.

Reviewer: ```suggestion
values and stores it to be called in next import. If we find that the hash value existed, meaning a function with the same parameter value set has been called before, for signed transaction this causes a function replacement if this call is set with a higher priority; and for unsigned transaction, this callback is simply ignored.
```

    service
      .keystore()
      .write()
      .insert_ephemeral_from_seed_by_type::<runtime::offchain_pallet::crypto::Pair>(
        &seed,
        runtime::offchain_pallet::KEY_TYPE,
      )
      .expect("Dev Seed should always succeed.");
  }
}
```

Then you will be able to sign transactions. This is good for **development only**.

### Option 2: Add an App Subkey via CLI

In a more realistic setting, after setting up your Substrate node, you can add a new app subkey via
the command line interface as follows.

```bash
# Generate a new account
$ subkey -s generate

# Submit a new key via RPC
$ curl -X POST -vk 'http://localhost:9933' -H "Content-Type:application/json;charset=utf-8" \
  -d '{
    "jsonrpc":2.0,
    "id":1,
    "method":"author_insertKey",
    "params": [
      "<YourKeyTypeId>",
      "<YourSeedPhrase>",
      "<YourPublicKey>"
    ]
  }'
```

A new key is now added in the local keystore.

## Signed Transactions

Now you are ready to to make a signed transaction from the off-chain worker. Go back to your pallet
in `my_offchain_worker.rs`.

```rust
decl_module! {
  pub struct Module<T: Trait> for enum Call where origin: T::Origin {
    // --snip--

>     // This is your regular on-chain extrinsics

Reviewer 2: This doesn't make sense to me.

      Ok(())
    }

    fn offchain_worker(block: T::BlockNumber) {
      // Here we specify the function to be called back on-chain in next block import.
      let call = Call::onchain_callback(block, b"hello world!".to_vec());
      T::SubmitSignedTransaction::submit_signed(call);
    }
  }
}
```

After having defined the on-chain callback function, in the off-chain worker you can specify
that function to be called back in the next block import phase. You then submit a signed transaction
to the node.

If you look at the implementation of `fn system::offchain::submit_signed` in the
[Substrate codebase](https://github.com/paritytech/substrate/blob/a98625501be68cc3084e666497c16b111741dded/frame/system/src/offchain.rs#L106-L115),
you will see it is calling the on-chain callback for each key in the local keystore. But since
you only have one key in the local keystore now, you are calling the function only once.

## Unsigned Transactions

With the following code, you are able to send an unsigned transaction back to the chain.

```rust
decl_module! {
  pub struct Module<T: Trait> for enum Call where origin: T::Origin {
    // --snip--

>     // This is your regular on-chain extrinsics

Reviewer 2: Ditto

    pub fn onchain_callback(_origin, _block: T::BlockNumber, input: Vec<u8>) -> dispatch::Result {
      debug::info!("{:?}", core::str::from_utf8(&input).unwrap());
      Ok(())
    }

    fn offchain_worker(block: T::BlockNumber) {
      // Here we specify the function to be called back on-chain in next block import.
      let call = Call::onchain_callback(block, b"hello world!".to_vec());
      T::SubmitUnsignedTransaction::submit_unsigned(call);
    }
  }
}
```

By default, all unsigned transactions are treated as invalid transactions. You need to add the
following code piece in `my_offchain_worker.rs` to explicitly allow some on-chain functions to call
unsigned transactions.

```rust
decl_module! {
  // --snip--
}

impl<T: Trait> Module<T> {
  // --snip--
}

#[allow(deprecated)]

> impl<T: Trait> support::unsigned::ValidateUnsigned for Module<T> {

Reviewer 2: This should be removed ASAP, including the paragraph that follows this code snippet. Unfortunately, it seems we don't actually have a code example of using `SignedExtension` to do this. At a minimum, we should like people to the conceptual-extrinsics doc where we talk about validate-unsigned.

I think we can probably merge now with the example so that we don't leave people hanging, but should really caution them with "don't make this a habit."

Author: Updated, and added links at the bottom of each section, and the last `Next Steps` section.

  type Call = Call<T>;

  fn validate_unsigned(call: &Self::Call) -> TransactionValidity {

    match call {
      Call::onchain_callback(block, input) => Ok(ValidTransaction {
        priority: 0,
        requires: vec![],
        provides: vec![(block, input).encode()],
        longevity: TransactionLongevity::max_value(),
        propagate: true,
      }),
      _ => InvalidTransaction::Call.into()
    }
  }
}
```

You see we add a `deprecated` attribute to prevent warning messages from being displayed. This is
because this part of the API is still in transition and will be updated in coming Substrate release.

## Parameters in On-Chain Callbacks

When making an on-chain callback, our implementation hashes the function name together with all

> of its parameter values. The callback will be stored and called during the next block next import.

Reviewer 3: ```suggestion
of its parameter values. The callback will be stored and called during the next block import.
```

      let call = Call::onchain_callback(block, b"hello world!".to_vec());
      T::SubmitUnsignedTransaction::submit_unsigned(call);
    }
  }
}
```

By default, all unsigned transactions are treated as invalid transactions. You need to add the
following code piece in `my_offchain_worker.rs` to explicitly allow submitting unsigned
transactions.

```rust
decl_module! {
  // --snip--
}

impl<T: Trait> Module<T> {
  // --snip--
}

#[allow(deprecated)]
impl<T: Trait> support::unsigned::ValidateUnsigned for Module<T> {
  type Call = Call<T>;

  fn validate_unsigned(call: &Self::Call) -> TransactionValidity {

    match call {
      Call::onchain_callback(block, input) => Ok(ValidTransaction {
        priority: 0,
        requires: vec![],
        provides: vec![(block, input).encode()],
        longevity: TransactionLongevity::max_value(),
        propagate: true,
      }),
      _ => InvalidTransaction::Call.into()
    }
  }
}
```

We add a `deprecated` attribute to prevent warning messages from being displayed. It is
because this part of the API is still in transition and will be updated in coming Substrate release.

> Please use this in caution for now.

Reviewer 2: ```suggestion
Please use this with caution for now.
```

      Ok(())
    }

    fn offchain_worker(block: T::BlockNumber) {
      // Here we specify the function to be callback on-chain in next block import
      let call = Call::onchain_callback(block, b"hello world!".to_vec());
      T::SubmitUnsignedTransaction::submit_unsigned(call);
    }
  }
}
```

By default, all unsigned transactions are treated as invalid transactions. You need to add the
following code piece in `offchain_pallet.rs` to explicitly allow some on-chain functions to be called
with unsigned transaction.

**runtime/src/offchain_pallet.rs**

```rust
decl_module! {
  // --snip--
}

impl<T: Trait> Module<T> {
  // --snip--
}

#[allow(deprecated)]
impl<T: Trait> support::unsigned::ValidateUnsigned for Module<T> {
  type Call = Call<T>;

  fn validate_unsigned(call: &Self::Call) -> TransactionValidity {

    match call {
      Call::onchain_callback(block, input) => Ok(ValidTransaction {
        priority: 0,
        requires: vec![],
        provides: vec![(block, input).encode()],
        longevity: TransactionLongevity::max_value(),
        propagate: true,
      }),
      _ => InvalidTransaction::Call.into()
    }
  }
}
```

> You see that we add a `deprecated` attribute to prevent warning message from display. This is because

Reviewer 3: ```suggestion
You see that we add a `deprecated` attribute to prevent warning messages from being displayed. This is because
```

this part of the API is still in transition and will be updated in coming Substrate release.

## Parameters in On-Chain Callbacks

> When making a on-chain callback, our implementation hashes the function name together with all

Author: Agree with the `an`. For the second modification, I am more on the side of using present tense.

parameter values and stores it to be called in next import. If we find that the hash value exists,
meaning a function with the same parameter value set has been called before, for signed transaction

> this causes a function replacement if called with a higher priority; for unsigned transaction, this

Reviewer 3: ```suggestion
, the function call will be replaced if it is called with a higher priority; for an unsigned transaction, this
```

callback is simply ignored.

> If your pallet is making on-chain callbacks regularly and you expect it will have duplicated

Reviewer 3: ```suggestion
If your pallet is making on-chain callbacks regularly and you expect it to have duplicate
```

> parameter value set every once in a while, you can always pass in an additional parameter of the

Reviewer 3: ```suggestion
sets of parameters every once in a while, you can always pass in an additional parameter of the
```

> current block number, that is being passed in from `fn offchain_worker()`. This number is going to

Reviewer 3: ```suggestion
current block number thorough the `fn offchain_worker()` function. This number is going to
```
